### PR TITLE
Support http server port reuse

### PR DESCRIPTION
**Pull request by zuston**
* Motivation for features / changes

ReusePort scenario: 
parent process occupies the port, then share the port through service such as ZooKeeper and [TonY](https://github.com/linkedin/TonY) (TensorFlow on Yarn), and then child process (TensorBoard process) reuse the port.

* Technical description of changes

* Screenshots of UI changes

* Detailed steps to verify changes work correctly (as executed by you)

* Alternate designs / implementations considered

### CODE REVIEW

**Comment by zuston** [comment]
@psybuzz @stephanwlee Please review it.

**Comment by stephanwlee** [comment]
Sorry about delayed review. The team is discussing the PR and is coming up with a response. We will definitely get back to you by next week. Thanks for understanding and your patience.

**Comment by zuston** [comment]
> Sorry about delayed review. The team is discussing the PR and is coming up with a response. We will definitely get back to you by next week. Thanks for understanding and your patience.

Thanks. Looking forward reply ~

**Review by wchargin — COMMENTED** [review]
Hi @zuston! Thanks for the pull request. The feature that you’re looking
for seems reasonable enough, but we’re hesitant to add more networking
configuration to TensorBoard proper, especially platform-specific code.

Our Python API enables you to provide a custom server class: you can use
the `tensorboard.program.TensorBoard` class and pass a `server_class`
argument to the constructor. The `program.TensorBoardServer` class
specifies the interface that it needs to implement: basically, your
class is given a WSGI app and has complete control over the server.

Have you tried this? If not, could you give it a shot and let us know if
it works for your use case?

**Comment by zuston** [comment]
Thanks reply @wchargin. 
It may not work. We provide users with a platform(on Hadoop Yarn) to deploy distributed TensorFlow. 
Therefore, it is hard to rewrite the user code and extend the implementation of a TensorBoard server.

Actually, some similiar changes are also on TensorFlow. like this https://github.com/tensorflow/tensorflow/pull/38705. 

Do you have some ideas on it? Thanks again.

**Comment by nfelt** [comment]
@zuston could you elaborate a little more about what constraint requires that you open the port in the parent process?  Do you have a link to some technical documentation about ZooKeeper or Yarn that explains how this works?  Without knowing more about your situation, it would seem more straightforward to just take whatever mechanism assigned the port number to the parent process and instead pass it directly through to TensorBoard.

**Comment by zuston** [comment]
我们在 Hadoop yarn 上进行 TensorFlow training, 但是 Hadoop yarn 原生不支持 TensorFlow。因此使用了 tony 框架，这个一个基于 Yarn 的扩展机制实现的框架。TonY is a framework to natively run deep learning jobs on Apache Hadoop. It currently supports TensorFlow, PyTorch, MXNet and Horovod.
简单说下 tony 的实现原理，就是向 Hadoop yarn 获取训练所需要的资源，比如机器节点、内存和 cpu。在获取到机器资源（tony task executor）后，tony 会在 task executor 获取角色需要的 grpc 端口和 tensorboard 启动端口(only in tensorflow chief role)。 
我们都知道，分布式 TensorFlow 的训练需要对每个角色指定 tf_config 环境配置。因此当上述操作完成后，TensorFlow 各个角色的 tf_config 就确定了。因此 tf_config 是由 tony 来动态决定的，且以 环境变量的形式注入到各个节点的运行时环境中。

我想这种机制与 kubeflow tf-operator 实现的原理是类似的。

回到这个 pull request 来说, 为什么需要端口复用？因为在 tony 的 task executor 在确定了一个可用端口之后，在启动用户的 trainning code 之前，

**Comment by zuston** [comment]
We use TensorFlow for model training on Hadoop Yarn, but Yarn does not natively support TensorFlow. Therefore, [the TonY framework](https://github.com/linkedin/TonY) is used, which is a framework to natively run deep learning jobs on Apache Hadoop. It currently supports TensorFlow, PyTorch, MXNet and Horovod.

Simply speaking, TonY's implementation principle is to obtain the resources needed for training from Hadoop Yarn, such as machine nodes, memory, and cpu. After obtaining the machine resource (TonY task executor), TonY will obtain the tensorflow grpc port and tensorboard startup port (only in tensorflow chief role) required by the role in the TonY task executor. But in this way, the tensorboard in the user code needs to specify the pre-allocated port of TonY when it is started. How to get tensorboard port in code? It can use the environment variable of TonY, [see the example for details](https://github.com/linkedin/TonY/blob/506bef815d71e87a7150819911535a6ed73e2167/tony-examples/mnist-tensorflow/mnist_distributed.py#L42).

As we all know that the distributed TensorFlow training needs to specify the tf_config environment configuration for each role. Therefore, when the above operations are completed, the tf_config of each role of TensorFlow is determined. Therefore, tf_config is dynamically determined by TonY and injected into the runtime environment of each node by environment variables. I think this mechanism is similar to kubeflow tf-operator (which provides a Kubernetes custom resource that makes it easy to run distributed or non-distributed TensorFlow jobs on Kubernetes.).

Back to this pull request, why do you need port reuse? Because after TonY's task executor determines an available port, it will release the port before starting the user's training code. Task executor will be subject to port conflict where port could be grabbed by other processes on the host before task process launches and after ports are closed.
Similar problems not only exist on tensorboard on TonY, but also on TensorFlow grpc service startup. TonY has now solved this problem through TensorFlow port reuse, [see here for details](https://github.com/linkedin/TonY/pull/456).
So I think it is necessary to solve the tensorboard port conflict problem in a similar way

reference:
https://arxiv.org/abs/1904.01631 (TonY paper)
https://github.com/linkedin/TonY/pull/456 (TonY related issue)

**Comment by zuston** [comment]
Any update on it? @nfelt @wchargin @stephanwlee

**Review by nfelt — CHANGES_REQUESTED** [review]
Thanks for the additional context about why you need the feature. We're willing to accept this PR but in addition to the comments left inline, could you also check that this behaves as expected on Windows?  (I don't think Windows has the `SO_REUSEPORT` option so we'll need code to handle that gracefully, which should be tested.)

Also, the tests are failing; you'll need to add several lines for `reuse_port` here:
https://github.com/tensorflow/tensorboard/blob/1597e6d39b89717b43b35c7871f62e30b1d0ade4/tensorboard/plugins/core/core_plugin_test.py#L46-L71

File: tensorboard/plugins/core/core_plugin.py [hunk 1]
```diff
@@ -341,6 +341,18 @@ def define_flags(self, parser):
...
 
+        parser.add_argument(
+            "--reuse_port",
+            metavar="BOOL",
+            # Custom str-to-bool converter since regular bool() doesn't work.
+            type=lambda v: {"true": True, "false": False}.get(v.lower(), v),
+            choices=[True, False],
+            default=False,
+            help="""\
+Enable http server port reuse. (default: %(default)s)\
```
[the remark this answers is not in the document]
Comment (zuston) [reply tensorboard/plugins/core/core_plugin.py]: Done

File: tensorboard/program.py [hunk 2]
```diff
@@ -682,6 +682,10 @@ def _get_wildcard_address(self, port):
         return fallback_address
 
     def server_bind(self):
+        # enable http server reuse port when set reuse_port param.
```
[the remark this answers is not in the document]
Comment (zuston) [reply tensorboard/program.py]: Done

File: tensorboard/program.py [hunk 3]
```diff
@@ -682,6 +682,10 @@ def _get_wildcard_address(self, port):
...
+        if self._flags.reuse_port:
```
[the remark this answers is not in the document]
Comment (zuston) [reply tensorboard/program.py:686]: Done

File: tensorboard/program.py [hunk 4]
```diff
@@ -682,6 +682,10 @@ def _get_wildcard_address(self, port):
...
+            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
```
[the remark this answers is not in the document]
Comment (zuston) [reply tensorboard/program.py:693]: Done.

File: tensorboard/program.py [hunk 5]
```diff
@@ -682,6 +682,10 @@ def _get_wildcard_address(self, port):
...
+
         """Override to enable IPV4 mapping for IPV6 sockets when desired.
```
[the remark this answers is not in the document]
Comment (zuston) [reply tensorboard/program.py]: Done

**Comment by zuston** [comment]
> Thanks for the additional context about why you need the feature. We're willing to accept this PR but in addition to the comments left inline, could you also check that this behaves as expected on Windows? (I don't think Windows has the SO_REUSEPORT option so we'll need code to handle that gracefully, which should be tested.)

With my colleague’s Windows PC, I have checked that it will throw exception as expected. @nfelt

**Comment by zuston** [comment]
Please review it. Thanks~ @nfelt @wchargin

**Review by nfelt — APPROVED** [review]
Thanks for the contribution, will merge.  Just FYI, most likely this won't be available in stable TensorBoard until our next release (2.5.0) which may be a month or two away, but it should be in nightly releases after tomorrow.

**Comment by zuston** [comment]
Thank you for your patience~. @nfelt

**Pull request merged**